### PR TITLE
Update to LTS-12

### PR DESCRIPTION
**Pull request by waddlaw**

### CODE REVIEW

**Comment by pbrisbin** [comment]
Thanks! Seems like we might need to get lucky on CI. Sometimes with resolver bumps it doesn't have the resources to complete a cache-less build. _Usually_ I can get something green on retries...

**Comment by pbrisbin** [comment]
It looks like we need to import `Semigroup` on compilers older than 8.4. If we want to keep supporting these non-latest versions we'll need `CPP` -- WDYT?

**Comment by waddlaw** [comment]
That makes sense. I will try support to `lts-10`, `lts-11`, `lts-12`.

File: .circleci/config.yml [hunk 1]
```diff
@@ -4,7 +4,8 @@ version: 2.0
 references:
   stack_build: &stack_build
     docker:
-      - image: fpco/stack-build:lts
+      # https://github.com/haskell-works/stack-build/blob/master/minimal/Dockerfile
+      - image: quay.io/haskell_works/stack-build-minimal
```
Comment (pbrisbin) [inline .circleci/config.yml]: Nice! I didn't know about this.

File: Makefile [hunk 2]
```diff
@@ -17,5 +18,10 @@ test:
 
 .PHONY: lint
 lint:
-	hlint .
-	weeder .
+	# hlint
+	curl -sL https://raw.github.com/ndmitchell/hlint/master/misc/travis.sh | sh -s -- --version
+	curl -sL https://raw.github.com/ndmitchell/hlint/master/misc/travis.sh | sh -s .
```
Comment (pbrisbin) [inline Makefile]: Can you describe the benefits of using this custom script over the simpler:

```
stack install --copy-compiler-tool hlint weeder
stack exec hlint
stack exec weeder
```

File: .circleci/config.yml (hunk 1 above)
Comment (waddlaw) [reply .circleci/config.yml]: Yes! very very fast! :)

File: Makefile (hunk 2 above)
Comment (waddlaw) [reply Makefile]: This script is able to use always latest version hlint (and weeder).

```
stack install --copy-compiler-tool hlint weeder
```

By above instruction is installed binaries of different version depends on that resolvers (`STACK_YAML` env).

Comment (pbrisbin) [reply Makefile]: Gotcha. I guess I could go either way: always latest sounds good, but using a tool appropriate for the compiler/dependencies versions under test could also be required in some edge-cases.

You've got it Green, so we can just go with this for now!

Comment (waddlaw) [reply Makefile]: Good point! Revert changes?

Comment (pbrisbin) [reply Makefile]: No, I think it's OK. We'll see how it goes.

Comment (waddlaw) [reply Makefile]: OK! :)
